// tropo/lib/daemon/daemon.js
/*global module,require*/
var domain = require('domain'),
    path = require('path'),
    flatiron = require('flatiron');

var cli = module.exports = flatiron.app;
cli.config.argv();

cli.config.defaults({
    provisioningServerPingIntervalMs: 30000
});

//
// Setup `tropo` to use `pkginfo` to expose version
//
require('pkginfo')(module, 'name', 'version');
var usage = ['Usage: tropod <command> [arg 1] [arg 2] [arg ...]'];

//
//Configure tropo to use `flatiron.plugins.cli`
//
cli.use(flatiron.plugins.cli, {
    version: true,
    source: path.join(__dirname, 'commands'),
    usage: usage,
    argv: {
        provisioningServerUrl: {
            description: 'URL of the provisioning server',
            string: true
        },
        provisioningServerUsername: {
            description: 'Your username for the provisioning server',
            string: true
        },
        provisioningServerPassword: {
            description: 'Your password for the provisioning server',
            string: true
        },
        version: {
            alias: 'v',
            description: 'print Tropo version and exit',
            string: true
        }
    }
});
