// tropo/test/ask.js
var Rayo = require('../lib/rayo')
var util = require('util');
var events = require('events');
var Tropo = require('../lib/tropo-impl');
var MockTransport = require('../lib/rayo-mock');
var fs = require('fs');
var XmlElement = require('../lib/utils').XmlElement;
var xml2js = require('xml2js');

var xmlParser = new xml2js.Parser();

var tropo = null;
var transport = null;

var defaultCallId = 'foocall';
var defaultConnectionId = 'fooconn';
var defaultTo = 'tel:+555';
var defaultFrom = 'tel:+666';

process.on('uncaughtException', function(err) {
  console.error(err.stack);
  process.exit(1);
});

module.exports = {

    setUp : function(callback) {
        var transportConfig = {
            type : 'mock'
        };
        var rayoConfig = {
        	callPingInterval : '5000'
        };

        tropo = new Tropo();
        tropo.debug = true;
        //tropo.addSubscriber('test@tropo', null, ['CALL_RING_LIST'], null, null, )
        
        // Initialize Tropo
        tropo.init({
            rayoTransportConfig : transportConfig,
            rayoConfig: rayoConfig,
            defaultPermissions: ['CALL_RING_LIST']
        });
        transport = tropo.mockTransport;
        callback();
    },
    tearDown: function(callback) {
    	
    	tropo.shutdown();
    	callback();
    },  
    'Ask succeeds' : function(test) {
        test.expect(16);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
            var result = call.connections[0].ask('Enter five digits', {choices:'[5 DIGITS]'});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'p', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getText(), 'Enter five digits', 'Unexpected grammar value ');                    
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
    'Ask with audio prompt succeeds' : function(test) {
        test.expect(16);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
            var result = call.connections[0].ask('http://www.myserver.com/a.mp3', {choices:'[5 DIGITS]'});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'audio', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].attrs['src'], 'http://www.myserver.com/a.mp3', 'Unexpected audio text');
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },

    'Ask with ssml prompt succeeds' : function(test) {
        test.expect(15);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
        	var ssml = '<speak><say-as interpret-as=\"ordinal\">100</say-as></speak>';                
            var result = call.connections[0].ask(ssml, {choices:'[5 DIGITS]'});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getText(), ssml, 'Unexpected Element Name');
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    

    'Ask with ssml prompt and attributes succeeds' : function(test) {
        test.expect(15);
        tropo.on('call:connected', function(event) {
            var subscriber = event.subscriber;
            var call = event.call
            process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
            });
            
            var ssml = '<speak version="0"><say-as interpret-as=\"ordinal\">100</say-as></speak>';                
            var result = call.connections[0].ask(ssml, {choices:'[5 DIGITS]'});
            process.nextTick(function() {
                transport.next(function(command, callback) {
                    console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getText(), ssml, 'Unexpected Element Name');
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
                    rayoEvent({
                        body: new XmlElement('complete')
                            .c('match').c('interpretation').t('12345').up()
                            .c('utterance').t('one two three four five').up().up(),
                        componentId: '123456'
                    });
                });
            });
            test.ok(result);
            test.deepEqual(result.grammars, ['[5 DIGITS]']);
            
            result.on('end', function(event) {
                test.ok(event);
                test.equals(result.result, '12345');
                test.equals(event.cause, 'match');
                test.equals(result.utterance, 'one two three four five');
                test.equals(event.call, call);
                test.equals(event.subscriber, subscriber);
                test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },    

    'Ask options succeeds' : function(test) {
        test.expect(39);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
            var options = {
                    mode: 'any',
                    terminator : '#',
                    recognizer : 'en_US',
                    initialTimeout : 5,
                    interDigitTimeout : 2,
                    sensitivity : 0.5,
                    minConfidence : 0.4,
                    maxSilence: 3,
                    choices: '[5 DIGITS]',
                    voice: 'veronica',
                    bargein: true,
                    attempts: 5
            };
        	
            var result = call.connections[0].ask('Enter five digits', options);
            test.equals(result.mode, 'any', 'Unexpected attribute value');
            test.equals(result.terminator, '#', 'Unexpected attribute value');
            test.equals(result.recognizer, 'en_US', 'Unexpected attribute value');
            test.equals(result.initialTimeout, '5', 'Unexpected attribute value');
            test.equals(result.interDigitTimeout, '2', 'Unexpected attribute value');
            test.equals(result.sensitivity, '0.5', 'Unexpected attribute value');
            test.equals(result.minConfidence, '0.4', 'Unexpected attribute value');
            test.equals(result.maxSilence, '3', 'Unexpected attribute value');
            test.equals(result.voice, 'veronica', 'Unexpected attribute value');
            test.equals(result.bargein, true, 'Unexpected attribute value');
            test.equals(result.bargeIn, true, 'Unexpected attribute value');
            test.equals(result.attempts, '5', 'Unexpected attribute value');
            
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'p', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getText(), 'Enter five digits', 'Unexpected grammar value ');                    
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    test.equals(command.body.attrs['mode'], 'any', 'Unexpected attribute value');
                    test.equals(command.body.attrs['terminator'], '#', 'Unexpected attribute value');
                    test.equals(command.body.attrs['recognizer'], 'en_US', 'Unexpected attribute value');
                    test.equals(command.body.attrs['initial-timeout'], '5000', 'Unexpected attribute value');
                    test.equals(command.body.attrs['inter-digit-timeout'], '2000', 'Unexpected attribute value');
                    test.equals(command.body.attrs['sensitivity'], '0.5', 'Unexpected attribute value');
                    test.equals(command.body.attrs['min-confidence'], '0.4', 'Unexpected attribute value');
                    test.equals(command.body.attrs['max-silence'], '3000', 'Unexpected attribute value');
                    test.equals(command.body.attrs['voice'], 'veronica', 'Unexpected attribute value');
                    test.equals(command.body.attrs['bargein'], 'true', 'Unexpected attribute value');
                    test.equals(command.body.attrs['attempts'], '5', 'Unexpected attribute value');
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
    'Ask with url grammar succeeds' : function(test) {
        test.expect(15);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
            var result = call.connections[0].ask('Enter five digits', {choices:'http://grammarserver.com/digits?min=1&max=10'});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'p', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getText(), 'Enter five digits', 'Unexpected grammar value ');                    
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['url'], 'http://grammarserver.com/digits?min=1&amp;max=10', 'Unexpected attribute value');
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['http://grammarserver.com/digits?min=1&max=10']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
    'Ask with SRGS grammar succeeds' : function(test) {
        test.expect(16);
    	var srgs = '<grammar version="1.0" xml:lang="en-US" ' +
        'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" ' + 
        'xsi:schemaLocation="http://www.w3.org/2001/06/grammar ' +
        'http://www.w3.org/TR/speech-grammar/grammar.xsd" ' +
        'xmlns="http://www.w3.org/2001/06/grammar">' + 
            										
           '<meta name="maxTime" content="4000"/>' +
           '<meta name="minSpeechDuration" content="4000"/>' +
           '<meta name="minVolume" content="10"/>' +
           '<meta name="finalSilence" content="2000"/>' +
           '<meta name="terminate" content="true"/>' +
           											
           '<ruleref uri="urn:xmpp:rayo:cpa:dtmf:1"/>' +
           '<ruleref uri="urn:xmpp:rayo:cpa:modem:1"/>' +
       '</grammar>';
        
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
            var grammar = new tropo.Grammar('application/srgs+xml', srgs);
            var result = call.connections[0].ask('Enter five digits', {choices:grammar});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'p', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getText(), 'Enter five digits', 'Unexpected grammar value ');                    
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/srgs+xml', 'Unexpected body value');
                    test.equals(command.body.children[1].children[0], '<![CDATA['+srgs+']]>', 'Unexpected body value');
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars[0].content, srgs);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },    
    
    'Ask with multiple grammars succeeds' : function(test) {
        test.expect(18);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
            var result = call.connections[0].ask('Enter five digits', {choices:['[5 DIGITS]','http://grammarserver.com/digits?min=1&max=10']});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'p', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getText(), 'Enter five digits', 'Unexpected grammar value ');                    
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    test.equals(command.body.children[2].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[2].attrs['url'], 'http://grammarserver.com/digits?min=1&amp;max=10', 'Unexpected attribute value');
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
	            	rayoEvent({
	            	    body: new XmlElement('complete')
	            			.c('match').c('interpretation').t('12345').up()
	            			.c('utterance').t('one two three four five').up().up(),
	            	    componentId: '123456'
	        		});
        	    });
        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]','http://grammarserver.com/digits?min=1&max=10']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(result.result, '12345');
    			test.equals(event.cause, 'match');
    			test.equals(result.utterance, 'one two three four five');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },

    'Ask stop succeeds' : function(test) {
        test.expect(12);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
        	
            var result = call.connections[0].ask('Enter five digits', {choices:'[5 DIGITS]'});
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log( command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
                    test.equals(command.body.children[0].getName(), 'prompt', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getName(), 'p', 'Unexpected Element Name');
                    test.equals(command.body.children[0].children[0].getText(), 'Enter five digits', 'Unexpected grammar value ');                    
                    test.equals(command.body.children[1].getName(), 'grammar', 'Unexpected Element Name');
                    test.equals(command.body.children[1].attrs['content-type'], 'application/grammar+voxeo', 'Unexpected attribute value');
                    test.equals(command.body.children[1].getText(), '[5 DIGITS]', 'Unexpected grammar value');                    
                    
                    callback(null, new XmlElement('ref', { id : '123456' }));
        	    });
                result.stop();
            	process.nextTick(function() {
	            	transport.next(function(command, callback) {
	            		test.equals(command.body.getName(), 'stop', 'Unexpected Element Name');
	            		
	                	rayoEvent({
	                	    body: new XmlElement('complete')
	                			.c('stop').up(),
	                	    componentId: '123456'
	            		})
	            	});
            	});

        	});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]']);
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(event.cause, 'stop');
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
    'Ask command fails if call has ended' : function(test) {
        test.expect(3);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect	                
        	});
        	            	
            var result = call.connections[0].ask('Hello', {choices:'[5 DIGITS]'});
    		test.ok(result);
    		test.deepEqual(result.grammars, ['[5 DIGITS]']);
        	process.nextTick(function() {
            	transport.next(function(command, callback) {
					callback(null, new XmlElement('ref', {id : '123456'}));
					
                    rayoEvent({
                        body: new XmlElement('end')
                    });
	                result.stop();
                });	            	
        	});
    		
    		result.on('error', function(event) {
				test.equals(event.error.message, 'Tropo Call [foocall], Connection [fooconn] has ended.');
        	    test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },  
    'Connection ask fails when connection is not connected' : function(test) {
    	
        test.expect(2);
        tropo.on('call:incoming', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call                
            var targets = ['sip:a@tropo.com','sip:b@tropo.com'];
		    var connection = call.connect(targets);

        	connection.on('error', function(error) {
			    test.ok(error);
    		});

            var result = connection.ask('Hello',{choices:'[5 DIGITS]'});
            test.ifError(result);
            test.done();		
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
    },     
    
    'Ask does not match' : function(test) {
        test.expect(8);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
            
            var result = call.connections[0].ask('Hello', {choices:'[5 DIGITS]'});
    		test.ok(result);
        	process.nextTick(function() {
                transport.next(function(command, callback) { 
            		console.log(command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
					callback(null, new XmlElement('ref', {id : '123456'}));
					
	            	rayoEvent({
	            	    body: new XmlElement('complete').c('nomatch').up(),
	            	    componentId: '123456'
	        		})
                });
        	});
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(event.cause, 'nomatch');
    			test.ifError(result.result);
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
    'Ask with no input' : function(test) {
        test.expect(8);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
            
            var result = call.connections[0].ask('Hello', {choices:'[5 DIGITS]'});
    		test.ok(result);
        	process.nextTick(function() {
                transport.next(function(command, callback) { 
            		console.log(command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
					callback(null, new XmlElement('ref', {id : '123456'}));
					
	            	rayoEvent({
	            	    body: new XmlElement('complete').c('noinput').up(),
	            	    componentId: '123456'
	        		})
                });
        	});
    		
    		result.on('end', function(event) {
    			test.ok(event);
    			test.equals(event.cause, 'noinput');
    			test.ifError(result.result);
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
    'Ask fails with error' : function(test) {
        test.expect(8);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
            
            var result = call.connections[0].ask('Hello', {choices:'[5 DIGITS]'});
    		test.ok(result);
        	process.nextTick(function() {
                transport.next(function(command, callback) { 
            		console.log(command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
					callback(null, new XmlElement('ref', {id : '123456'}));
					
	            	rayoEvent({
	            	    body: new XmlElement('complete').c('error').t('Something bad happened').up(),
	            	    componentId: '123456'
	        		})
                })
        	})
            
    		result.on('end', function(event) {
    			test.ok(false, "this event is not expected");
    		});
    		
    		result.on('error', function(event) {
    			test.ok(event.error);
    			test.ifError(result.result);
    			test.equals(event.error.message, 'Something bad happened');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    'Ask ends on hangup' : function(test) {
        test.expect(8);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
            
            var result = call.connections[0].ask('Hello', {choices:'[5 DIGITS]'});
    		test.ok(result);
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log(command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
					callback(null, new XmlElement('ref', {id : '123456'}));
					
	            	rayoEvent({
	            	    body: new XmlElement('complete').c('hangup').up(),
	            	    componentId: '123456'
	        		})
                })
        	})
    		result.on('end', function(event) {
    			test.ok(event);
    			test.ifError(result.result);
    			test.equals(event.cause, 'hangup');
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    'Ask fires start event' : function(test) {
        test.expect(5);
        tropo.on('call:connected', function(event) {
        	var subscriber = event.subscriber;
        	var call = event.call
        	process.nextTick(function() {
                transport.next(function(command, callback) { }); // connect
        	});
            
            var result = call.connections[0].ask('Hello', {choices:'[5 DIGITS]'});
    		test.ok(result);
        	process.nextTick(function() {
                transport.next(function(command, callback) {
            		console.log(command.body);
                    test.equals(command.body.getName(), 'ask', 'Unexpected Element Name');
					callback(null, new XmlElement('ref', {id : '123456'}));
                })
        	})
    		result.on('start', function(event) {
    			test.equals(event.call, call);
    			test.equals(event.subscriber, subscriber);
    			test.equals(event.ask, result);
                test.done();
            })
        });
        newCall({subscriber: defaultTo, permissions: ['CALL_RING_LIST']});
        
        var blegId = 'o1';
        rayoEvent({
            body: new XmlElement('announce').c('joining', {'call-id':blegId, to:defaultTo}).up()
        });
        rayoEvent({
            body: new XmlElement('joined'),
            connectionId : blegId
        });
    },
    
};

function rayoEvent(event) {
    var callId = defaultCallId;
    var connectionId = defaultConnectionId;
    var componentId = null;
    if (event.callId) {
        callId = event.callId;
    } 
    if (event.connectionId) {
    	connectionId = event.connectionId;
    }
    if (event.componentId) {
        componentId = event.componentId;
    } 
    
    transport.emit('callEvent', {
        callId : callId,
        connectionId : connectionId,
        componentId : componentId,
        body : event.body
    });
};

function newCall(options) {
    var callId = defaultCallId;
    var connectionId = defaultConnectionId;
    var direction = 'in';
    if (options) {
        if (options.callId) {
            callId = options.callId;
        }
        if (options.connectionId) {
        	connectionId = options.connectionId;
        }
        if (options.direction) {
            direction = options.direction;
        }
    }
    var body = new XmlElement('offer', {
        to : defaultTo,
        from : defaultFrom,
        'direction' : direction,
    });

	var subscription = body.c('subscription');
	subscription.c('address').t(defaultTo);
	if (options.permissions) {
		for (var permission in options.permissions) {
			subscription.c('permission', {'name':options.permissions[permission]})
		}
	}	
    
    transport.emit('subscribe', {address: defaultTo, permissions: ['CALL_RING_LIST']});
    transport.emit('callEvent', {
        callId : callId,
        connectionId : connectionId,
        body : body
    });
};
